### ISSUE TITLE
Kernel heap is slightly thread-unsafe

### ISSUE THREAD

**Issue by jeremydavis519** [jeremydavis519/phoenix#1]
It's extremely rare, but the fuzzing test for thread-safety in `memory::phys::heap` found an error:

```
---- phys::heap::tests::many_threads::should_keep_the_heap_valid stdout ----
Static master block: [0x00007f0f_3761a640, 0x00007f0f_3761b680)
PRNG seed: 1615482185842515200
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `15276209936040722432`,
 right: `14987979559889010688`', crates/memory/src/phys/heap/node.rs:314:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at 'expected an earlier node when removing a node from the heap', crates/memory/src/phys/heap/node.rs:285:25
panic occurred in a child thread: assertion failed: `(left == right)`
  left: `15276209936040722432`,
 right: `14987979559889010688`
panic occurred in a child thread: expected an earlier node when removing a node from the heap
thread 'phys::heap::tests::many_threads::should_keep_the_heap_valid' panicked at 'number of panicking child threads: 2', crates/memory/src/phys/heap/mod.rs:933:13
```

This will be almost impossible to reproduce, since it depends on how the host OS decides to schedule the threads.

I put a TODO comment in `collections::atomic::linked_list` that may indicate how we can simplify the heap data structure by getting rid of guard nodes. The process of simplification may fix this bug.

**Comment by jeremydavis519** [comment]
This might be fixed by bf61f20. Closing this issue for now, but it should be reopened it if the bug resurfaces.

**Comment by jeremydavis519** [comment]
Nope, not fixed. Here's another instance of the error:

```
---- phys::heap::tests::many_threads::should_keep_the_heap_valid stdout ----
Static master block: [0x00007fc4_dea0f640, 0x00007fc4_dea10680)
PRNG seed: 1616704105895575200
Validating node @ 0x00007fc4dea0fa80: [0x00007fc4_de90f040, 0x00007fc4_de910080)
Validating node @ 0x00007fc4dea0f800: [0x00007fc4_de910080, 0x00007fc4_de9110c0)
Validating node @ 0x00007fc4dea0f9c0: [0x00007fc4_de9110c0, 0x00007fc4_de912100)
Validating node @ 0x00007fc4dea0f7c0: [0x00007fc4_de912100, 0x00007fc4_de913140)
Validating node @ 0x00007fc4dea0fac0: [0x00007fc4_de913140, 0x00007fc4_de914180)
Validating node @ 0x00007fc4dea0f940: [0x00007fc4_de914180, 0x00007fc4_de9151c0)
Validating node @ 0x00007fc4dea0fc40: [0x00007fc4_de9151c0, 0x00007fc4_de916200)
Validating node @ 0x00007fc4dea0fd40: [0x00007fc4_de916200, 0x00007fc4_de917240)
Validating node @ 0x00007fc4de914240: [0x00007fc4_de917240, 0x00007fc4_de918280)
Validating node @ 0x00007fc4dea0fe80: [0x00007fc4_de918280, 0x00007fc4_de9192c0)
Validating node @ 0x00007fc4dea0f880: [0x00007fc4_de9192c0, 0x00007fc4_de91a300)
Validating node @ 0x00007fc4dea0fec0: [0x00007fc4_de91a300, 0x00007fc4_de91b340)
Validating node @ 0x00007fc4dea10300: [0x00007fc4_de91b340, 0x00007fc4_de91c380)
Validating node @ 0x00007fc4de90f700: [0x00007fc4_de91c380, 0x00007fc4_de91d3c0)
Validating node @ 0x00007fc4de913d80: [0x00007fc4_de91d3c0, 0x00007fc4_de91e400)
Validating node @ 0x00007fc4de912f80: [0x00007fc4_de91e400, 0x00007fc4_de91f440)
Validating node @ 0x00007fc4de915f00: [0x00007fc4_de91f440, 0x00007fc4_de920480)
Validating node @ 0x00007fc4de911ec0: [0x00007fc4_de920480, 0x00007fc4_de9214c0)
Validating node @ 0x00007fc4de914c00: [0x00007fc4_de9214c0, 0x00007fc4_de922500)
Validating node @ 0x00007fc4de915100: [0x00007fc4_de922500, 0x00007fc4_de923540)
Validating node @ 0x00007fc4dea10080: [0x00007fc4_de923540, 0x00007fc4_de924580)
Validating node @ 0x00007fc4de914bc0: [0x00007fc4_de924580, 0x00007fc4_de9255c0)
Validating node @ 0x00007fc4de914b80: [0x00007fc4_de9255c0, 0x00007fc4_de926600)
Validating node @ 0x00007fc4de9106c0: [0x00007fc4_de926600, 0x00007fc4_de927640)
Validating node @ 0x00007fc4de910a80: [0x00007fc4_de927640, 0x00007fc4_de928680)
Validating node @ 0x00007fc4de90f880: [0x00007fc4_de928680, 0x00007fc4_de9296c0)
Validating node @ 0x00007fc4de916ec0: [0x00007fc4_de9296c0, 0x00007fc4_de92a700)
Validating node @ 0x00007fc4de914900: [0x00007fc4_de92a700, 0x00007fc4_de92b740)
Validating node @ 0x00007fc4de919b80: [0x00007fc4_de92b740, 0x00007fc4_de92c780)
Validating node @ 0x00007fc4de916100: [0x00007fc4_de92c780, 0x00007fc4_de92d7c0)
Validating node @ 0x00007fc4de910680: [0x00007fc4_de92d7c0, 0x00007fc4_de92e800)
Validating node @ 0x00007fc4de915ac0: [0x00007fc4_de92e800, 0x00007fc4_de92f840)
Validating node @ 0x00007fc4de912100: [0x00007fc4_de92f840, 0x00007fc4_de930880)
Validating node @ 0x00007fc4de912600: [0x00007fc4_de930880, 0x00007fc4_de9318c0)
Validating node @ 0x00007fc4dea10000: [0x00007fc4_de9320c0, 0x00007fc4_de933100)
Validating node @ 0x00007fc4dea0fb80: [0x00007fc4_de933100, 0x00007fc4_de934140)
Validating node @ 0x00007fc4dea10100: [0x00007fc4_de934140, 0x00007fc4_de935180)
Validating node @ 0x00007fc4dea10280: [0x00007fc4_de935180, 0x00007fc4_de9361c0)
Validating node @ 0x00007fc4dea104c0: [0x00007fc4_de9361c0, 0x00007fc4_de937200)
Validating node @ 0x00007fc4dea0f700: [0x00007fc4_de937200, 0x00007fc4_de938240)
Validating node @ 0x00007fc4dea10540: [0x00007fc4_de938240, 0x00007fc4_de939280)
Validating node @ 0x00007fc4dea10480: [0x00007fc4_de939280, 0x00007fc4_de93a2c0)
Validating node @ 0x00007fc4de911440: [0x00007fc4_de93a2c0, 0x00007fc4_de93b300)
Validating node @ 0x00007fc4de9105c0: [0x00007fc4_de93b300, 0x00007fc4_de93c340)
Validating node @ 0x00007fc4de911740: [0x00007fc4_de93c340, 0x00007fc4_de93d380)
Validating node @ 0x00007fc4de912400: [0x00007fc4_de93d380, 0x00007fc4_de93e3c0)
Validating node @ 0x00007fc4de912300: [0x00007fc4_de93e3c0, 0x00007fc4_de93f400)
Validating node @ 0x00007fc4de911e00: [0x00007fc4_de93f400, 0x00007fc4_de940440)
Validating node @ 0x00007fc4de911fc0: [0x00007fc4_de940440, 0x00007fc4_de941480)
Validating node @ 0x00007fc4de9133c0: [0x00007fc4_de941480, 0x00007fc4_de9424c0)
Validating node @ 0x00007fc4de915980: [0x00007fc4_de9424c0, 0x00007fc4_de943500)
Validating node @ 0x00007fc4de910200: [0x00007fc4_de943500, 0x00007fc4_de944540)
Validating node @ 0x00007fc4de911bc0: [0x00007fc4_de944540, 0x00007fc4_de945580)
Validating node @ 0x00007fc4de912640: [0x00007fc4_de945580, 0x00007fc4_de9465c0)
Validating node @ 0x00007fc4de911700: [0x00007fc4_de9465c0, 0x00007fc4_de947600)
Validating node @ 0x00007fc4de911dc0: [0x00007fc4_de947600, 0x00007fc4_de948640)
Validating node @ 0x00007fc4de913a40: [0x00007fc4_de948640, 0x00007fc4_de949680)
Validating node @ 0x00007fc4de915d80: [0x00007fc4_de949680, 0x00007fc4_de94a6c0)
Validating node @ 0x00007fc4de90fdc0: [0x00007fc4_de94a6c0, 0x00007fc4_de94b700)
Validating node @ 0x00007fc4de912c80: [0x00007fc4_de94bc40, 0x00007fc4_de94cc80)
Validating node @ 0x00007fc4de914940: [0x00007fc4_de94cc80, 0x00007fc4_de94dcc0)
Validating node @ 0x00007fc4de910e00: [0x00007fc4_de94dcc0, 0x00007fc4_de94ed00)
Validating node @ 0x00007fc4de915e40: [0x00007fc4_de94ed00, 0x00007fc4_de94fd40)
Validating node @ 0x00007fc4de912dc0: [0x00007fc4_de94fd40, 0x00007fc4_de950d80)
Validating node @ 0x00007fc4de90fd40: [0x00007fc4_de950d80, 0x00007fc4_de951dc0)
Validating node @ 0x00007fc4de918b00: [0x00007fc4_de951dc0, 0x00007fc4_de952e00)
Validating node @ 0x00007fc4de918bc0: [0x00007fc4_de952e00, 0x00007fc4_de953e40)
Validating node @ 0x00007fc4de912080: [0x00007fc4_de953e40, 0x00007fc4_de954e80)
Validating node @ 0x00007fc4de916f80: [0x00007fc4_de954e80, 0x00007fc4_de955ec0)
Validating node @ 0x00007fc4de918100: [0x00007fc4_de955ec0, 0x00007fc4_de956f00)
Validating node @ 0x00007fc4de9157c0: [0x00007fc4_de956f00, 0x00007fc4_de957f40)
Validating node @ 0x00007fc4de917a80: [0x00007fc4_de957f40, 0x00007fc4_de958f80)
Validating node @ 0x00007fc4de916e80: [0x00007fc4_de958f80, 0x00007fc4_de959fc0)
Validating node @ 0x00007fc4de9184c0: [0x00007fc4_de959fc0, 0x00007fc4_de95b000)
Validating node @ 0x00007fc4de917b00: [0x00007fc4_de95b000, 0x00007fc4_de95c040)
Validating node @ 0x00007fc4de9192c0: [0x00007fc4_de95c040, 0x00007fc4_de95d080)
Validating node @ 0x00007fc4de914e40: [0x00007fc4_de95d7c0, 0x00007fc4_de95e800)
Validating node @ 0x00007fc4de90fe80: [0x00007fc4_de95e800, 0x00007fc4_de95f840)
Validating node @ 0x00007fc4de9138c0: [0x00007fc4_de95f840, 0x00007fc4_de960880)
Validating node @ 0x00007fc4de90f840: [0x00007fc4_de960880, 0x00007fc4_de9618c0)
Validating node @ 0x00007fc4dea105c0: [0x00007fc4_de9618c0, 0x00007fc4_de962900)
Validating node @ 0x00007fc4de90ff80: [0x00007fc4_de962900, 0x00007fc4_de963940)
Validating node @ 0x00007fc4de915b40: [0x00007fc4_de963940, 0x00007fc4_de964980)
Validating node @ 0x00007fc4de912140: [0x00007fc4_de964980, 0x00007fc4_de9659c0)
Validating node @ 0x00007fc4de913d40: [0x00007fc4_de9659c0, 0x00007fc4_de966a00)
Validating node @ 0x00007fc4de915580: [0x00007fc4_de966a00, 0x00007fc4_de967a40)
Validating node @ 0x00007fc4de916240: [0x00007fc4_de968940, 0x00007fc4_de969980)
Validating node @ 0x00007fc4de912e80: [0x00007fc4_de969980, 0x00007fc4_de96a9c0)
Validating node @ 0x00007fc4de916740: [0x00007fc4_de96a9c0, 0x00007fc4_de96ba00)
Validating node @ 0x00007fc4de916c80: [0x00007fc4_de96ba00, 0x00007fc4_de96ca40)
Validating node @ 0x00007fc4de916480: [0x00007fc4_de96ca40, 0x00007fc4_de96da80)
Validating node @ 0x00007fc4de915540: [0x00007fc4_de96da80, 0x00007fc4_de96eac0)
Validating node @ 0x00007fc4de910640: [0x00007fc4_de96eac0, 0x00007fc4_de96fb00)
Validating node @ 0x00007fc4de9159c0: [0x00007fc4_de96fb00, 0x00007fc4_de970b40)
Validating node @ 0x00007fc4de910a40: [0x00007fc4_de970e00, 0x00007fc4_de971e40)
Validating node @ 0x00007fc4de912000: [0x00007fc4_de971e40, 0x00007fc4_de972e80)
Validating node @ 0x00007fc4de913d00: [0x00007fc4_de972e80, 0x00007fc4_de973ec0)
Validating node @ 0x00007fc4de914d40: [0x00007fc4_de973ac0, 0x00007fc4_de974b00)
thread 'phys::heap::tests::many_threads::should_keep_the_heap_valid' panicked at '(0x7fc4de973ac0, 0x1040) is after (0x7fc4de972e80, 0x1040) but should be before it', crates/memory/src/phys/heap/mod.rs:1043:13
```

**Comment by jeremydavis519** [comment]
Closing because this is most likely fixed at least as of 8ee14c02d05b84b3feb69e14664aa4dbb8e510b2. That commit and the ones leading up to it greatly simplified the heap and the logic that keeps it consistent.
